File: android/liveness/src/main/java/com/liveness/detection/LivenessStep.kt
package com.liveness.detection

/**
 * Canonical liveness challenge step.
 *
 * `key` is the short lowercase identifier used in configuration (matching the
 * web SDK / sound-key form). `label` is the display string surfaced to the user.
 */
enum class LivenessStep(val index: Int, val key: String, val label: String) {
  TURN_LEFT(0,  "left",  "Turn your head LEFT"),
  BLINK    (1,  "blink", "Blink"),
  TURN_RIGHT(2, "right", "Turn your head RIGHT"),
  NOD      (3,  "nod",   "Nod your head"),
  MOUTH    (4,  "mouth", "Open your mouth");

  companion object {
    val ordered = listOf(TURN_LEFT, BLINK, TURN_RIGHT, NOD, MOUTH)

    /**
     * Resolve a caller-supplied list of step keys (e.g. `["nod", "blink"]`) into the
     * matching [LivenessStep]s. Empty input, all-invalid input, or `null` falls back
     * to all 5 steps in canonical order. Duplicates are dropped (first occurrence wins).
     */
    fun resolve(keys: List<String>?): List<LivenessStep> {
      if (keys.isNullOrEmpty()) return ordered
      val byKey = ordered.associateBy { it.key }
      val seen = HashSet<LivenessStep>()
      val out = ArrayList<LivenessStep>()
      for (k in keys) {
        val step = byKey[k.lowercase()] ?: continue
        if (seen.add(step)) out.add(step)
      }
      return if (out.isEmpty()) ordered else out
    }
  }
}

File: ios/LivenessDetection/LivenessStep.swift
import Foundation

public enum LivenessStep: Int, CaseIterable {
  case turnLeft = 0
  case blink = 1
  case turnRight = 2
  case nod = 3
  case mouth = 4

  /// Display string surfaced to the user.
  public var label: String {
    switch self {
    case .turnLeft:  return "Turn your head LEFT"
    case .blink:     return "Blink"
    case .turnRight: return "Turn your head RIGHT"
    case .nod:       return "Nod your head"
    case .mouth:     return "Open your mouth"
    }
  }

  /// Short lowercase identifier used in configuration (matches the web SDK / sound-key form).
  public var key: String {
    switch self {
    case .turnLeft:  return "left"
    case .blink:     return "blink"
    case .turnRight: return "right"
    case .nod:       return "nod"
    case .mouth:     return "mouth"
    }
  }

  /// Resolve a caller-supplied list of step keys into the matching steps.
  /// Empty input, all-invalid input or `nil` falls back to all 5 in canonical order.
  /// Duplicates are dropped (first occurrence wins).
  public static func resolve(keys: [String]?) -> [LivenessStep] {
    guard let keys = keys, !keys.isEmpty else { return Array(LivenessStep.allCases) }
    let byKey = Dictionary(uniqueKeysWithValues: LivenessStep.allCases.map { ($0.key, $0) })
    var seen = Set<LivenessStep>()
    var out: [LivenessStep] = []
    for k in keys {
      if let step = byKey[k.lowercased()], !seen.contains(step) {
        seen.insert(step)
        out.append(step)
      }
    }
    return out.isEmpty ? Array(LivenessStep.allCases) : out
  }
}
